### PR TITLE
boards/stm32: add composite device configurations

### PR DESCRIPTION
**Pull request by raiden00pl**
## Summary

- stm32f4discovery: add composite device configuration
- stm32f4discovery/stm32_usbmsc.c: fix SD card initialization
- stm32f4discovery/stm32_mmcsd.c: remove redundant SCK pin configuration, it's already selected in board.h
- nucleo-144: add composite device configuration
- nucleo-h743zi: add composite device configuration

## Testing
Added configuration

### CODE REVIEW

**Comment by davids5** [comment]
@raiden00pl Did you rest this on the all the effected HW?

**Comment by raiden00pl** [comment]
Yes, they were all tested. They are based on the same IP core (dwc2 from Synopsys) and I hoped that one stm32 would work as a composite device. Unfortunately, the main problems were in USB drivers but at least some new configs were created.

**Review by davids5 — APPROVED** [review]
Thank you @raiden00pl

File: boards/arm/stm32h7/nucleo-h743zi/src/stm32_composite.c [hunk 1]
```diff
@@ -0,0 +1,347 @@
...
+  dev[dev_idx].devinfo.epno[CDCACM_EP_INTIN_IDX] = epin++;
+  dev[dev_idx].devinfo.epno[CDCACM_EP_BULKIN_IDX] = epin++;
+  dev[dev_idx].devinfo.epno[CDCACM_EP_BULKOUT_IDX] = epout++;
+
+  /* Count up the base numbers */
+
+  ifnobase += dev[dev_idx].devinfo.ninterfaces;
+  strbase  += dev[dev_idx].devinfo.nstrings;
+
+  dev_idx += 1;
```
Comment (pkarashchenko) [inline boards/arm/stm32h7/nucleo-h743zi/src/stm32_composite.c:253]: Nitpick:

```suggestion
  dev_idx++;
```

File: boards/arm/stm32h7/nucleo-h743zi/src/stm32_composite.c [hunk 2]
```diff
@@ -0,0 +1,347 @@
...
+  dev[dev_idx].devinfo.epno[RNDIS_EP_INTIN_IDX] = epin++;
+  dev[dev_idx].devinfo.epno[RNDIS_EP_BULKIN_IDX] = epin++;
+  dev[dev_idx].devinfo.epno[RNDIS_EP_BULKOUT_IDX] = epout++;
+
+  /* Count up the base numbers */
+
+  ifnobase += dev[dev_idx].devinfo.ninterfaces;
+  strbase += dev[dev_idx].devinfo.nstrings;
+
+  dev_idx += 1;
```
Comment (pkarashchenko) [inline boards/arm/stm32h7/nucleo-h743zi/src/stm32_composite.c:214]:
```suggestion
  dev_idx++;
```

File: boards/arm/stm32h7/nucleo-h743zi/src/stm32_composite.c [hunk 3]
```diff
@@ -0,0 +1,347 @@
...
+
+  dev[dev_idx].devinfo.epno[USBMSC_EP_BULKIN_IDX] = epin++;
+  dev[dev_idx].devinfo.epno[USBMSC_EP_BULKOUT_IDX] = epout++;
+
+  /* Count up the base numbers */
+
+  ifnobase += dev[dev_idx].devinfo.ninterfaces;
+  strbase  += dev[dev_idx].devinfo.nstrings;
+
+  dev_idx += 1;
```
Comment (pkarashchenko) [inline boards/arm/stm32h7/nucleo-h743zi/src/stm32_composite.c:291]:
```suggestion
  dev_idx++;
```

**Review by pkarashchenko — APPROVED** [review]
Nothing that blocks from merging

**Comment by xiaoxiang781216** [comment]
@raiden00pl please fix the conflict.

**Comment by raiden00pl** [comment]
@xiaoxiang781216 conflict fixed

File: boards/arm/stm32f7/nucleo-144/src/stm32_composite.c [hunk 4]
```diff
@@ -0,0 +1,347 @@
...
+ *   A non-NULL handle value is returned on success.  NULL is returned on
+ *   any failure.
+ *
+ ****************************************************************************/
+
+static void *board_composite0_connect(int port)
+{
+  struct composite_devdesc_s dev[COMPOSITE0_DEV];
+  int ifnobase = 0;
+  int strbase  = COMPOSITE_NSTRIDS;
```
Comment (pkarashchenko) [inline boards/arm/stm32f7/nucleo-144/src/stm32_composite.c:180]:
```suggestion
  int strbase = COMPOSITE_NSTRIDS;
```

File: boards/arm/stm32h7/nucleo-h743zi/src/stm32_composite.c [hunk 5]
```diff
@@ -0,0 +1,347 @@
...
+ *   A non-NULL handle value is returned on success.  NULL is returned on
+ *   any failure.
+ *
+ ****************************************************************************/
+
+static void *board_composite0_connect(int port)
+{
+  struct composite_devdesc_s dev[COMPOSITE0_DEV];
+  int ifnobase = 0;
+  int strbase  = COMPOSITE_NSTRIDS;
```
Comment (pkarashchenko) [inline boards/arm/stm32h7/nucleo-h743zi/src/stm32_composite.c:180]:
```suggestion
  int strbase = COMPOSITE_NSTRIDS;
```

File: boards/arm/stm32h7/nucleo-h743zi/src/nucleo-h743zi.h [hunk 6]
```diff
@@ -196,6 +196,16 @@
                             GPIO_OUTPUT_CLEAR | GPIO_PORTF | GPIO_PIN12)
 #define GPIO_NRF24L01_IRQ  (GPIO_INPUT | GPIO_FLOAT | GPIO_PORTD | GPIO_PIN15)
 
+/* MMC/SD
+ * CS  - PD15 (D9)
+ * NCD - PF12 (D8)
+ */
+
+#define GPIO_MMCSD_CS    (GPIO_OUTPUT | GPIO_PUSHPULL | GPIO_SPEED_50MHz | \
+                           GPIO_OUTPUT_SET | GPIO_PORTD | GPIO_PIN15)
```
Comment (davids5) [inline boards/arm/stm32h7/nucleo-h743zi/src/nucleo-h743zi.h:205]:
```suggestion
                          GPIO_OUTPUT_SET | GPIO_PORTD | GPIO_PIN15)
```

**Review by davids5 — CHANGES_REQUESTED** [review]
@raiden00pl Looks good but some minor changes.

**Comment by xiaoxiang781216** [comment]
@raiden00pl will you fix the remaining comment from @pkarashchenko ?

**Comment by raiden00pl** [comment]
@xiaoxiang781216  I prefer to keep the code as it is now.

**Pull request merged**